### ISSUE TITLE
upgrade flake: timeout waiting old image pod deleted

### ISSUE THREAD

**Issue by hasbro17** [coreos/etcd-operator#1076]
Currently we use a timeout of 30s in `WaitPodsWithImageDeleted()` in etcd operator upgrade:

```go
f.KubeCli.AppsV1beta1().Deployments(f.KubeNS).Patch(d.Name, types.StrategicMergePatchType, patchData)
...
e2eutil.WaitPodsWithImageDeleted(f.KubeCli, f.KubeNS, f.OldImage, 30*time.Second, lo)
```

There has been at least 1 instance of this causing a timeout failure in the upgrade test:
```
cluster: "jenkins"
OPERATOR_IMAGE: gcr.io/coreos-k8s-scale-testing/etcd-operator:21355118f9527430d35249b8234b2aeb76ae3bd9
TEST_NAMESPACE: e2e-etcd-operator-upgrade-19

--- FAIL: TestResize (90.58s)
	util.go:91: 2017-05-08 16:10:44.859024118 +0000 UTC created etcd cluster: upgrade-test-1dkfg
	util.go:91: 2017-05-08 16:10:44.89568783 +0000 UTC waiting size (3), etcd pods: names ([]), nodes ([])
	util.go:91: 2017-05-08 16:10:54.896758238 +0000 UTC waiting size (3), etcd pods: names ([upgrade-test-1dkfg-0000]), nodes ([gke-jenkins-default-pool-8bb205c4-rjb4])
	util.go:91: 2017-05-08 16:11:04.896635429 +0000 UTC waiting size (3), etcd pods: names ([upgrade-test-1dkfg-0000 upgrade-test-1dkfg-0001]), nodes ([gke-jenkins-default-pool-8bb205c4-rjb4 gke-jenkins-default-pool-8bb205c4-k5tk])
	util.go:91: 2017-05-08 16:11:14.897175996 +0000 UTC waiting size (3), etcd pods: names ([upgrade-test-1dkfg-0000 upgrade-test-1dkfg-0001 upgrade-test-1dkfg-0002]), nodes ([gke-jenkins-default-pool-8bb205c4-rjb4 gke-jenkins-default-pool-8bb205c4-k5tk gke-jenkins-default-pool-8bb205c4-xmgs])
	upgrade_test.go:59: still failing after 6 retries
	util.go:91: 2017-05-08 16:12:05.088677594 +0000 UTC waiting pod (upgrade-test-1dkfg-0000) to be deleted.
	wait_util.go:102: pod (upgrade-test-1dkfg-0000) status.phase is (Running): init container status:
		container status:
	upgrade_test.go:50: fail to wait pods deleted: still failing after 6 retries
FAIL
FAIL	github.com/coreos/etcd-operator/test/e2e/upgradetest	408.195s
Fail! ===
```
We should keep an eye on this if this fails repeatedly in the future.

**Comment by hongchaodeng** [comment]
Another recent failure:
https://jenkins-etcd.prod.coreos.systems/view/operator/job/etcd-operator-upgrade/30/

Based on our initial analysis, it seemed like a timing issue.

We should consider measuring, collecting how much each step takes, e.g. 99% . And then better define timeout.

**Comment by hasbro17** [comment]
Currently when an upgrade test creates an operator, and then creates a cluster of size 3, it can take up to 60s for the last etcd pod to be created:
```
POD TIMES:
util.go:91: 2017-05-31 20:57:26.371054714 +0000 UTC pod: (upgrade-test-recovery-p091g-0000) took (20.038399589s)
util.go:91: 2017-05-31 20:57:26.371061778 +0000 UTC pod: (upgrade-test-recovery-p091g-0001) took (30.039262606s)
util.go:91: 2017-05-31 20:57:26.371066701 +0000 UTC pod: (upgrade-test-recovery-p091g-0002) took (50.040822013s)
```
Because of this, the tests that use a timeout of 60s for `WaitUntilSizeReached()` can fail occasionally.

We can either increase the timeout, or possibly find out why the first pod takes 20s to create. 
This could be because the etcd-operator is not fully running by the time we call `WaitUntilSizeReached()`. So the other option is to wait until the etcd-operator is running before creating the cluster.

**Comment by hongchaodeng** [comment]
@hasbro17 
Can you write a minimal step to reproduce it and paste here?

**Comment by hongchaodeng** [comment]
This is reproducible.

**Comment by hongchaodeng** [comment]
I found something suspiciously long.
It took almost 20s from
```
"Go OS/Arch: linux/amd64"
```
to
```
finding existing clusters...
```

I suspect most of time is spent on this func:
```Go
func (c *Controller) createTPR() error {
	tpr := &v1beta1extensions.ThirdPartyResource{
		ObjectMeta: metav1.ObjectMeta{
			Name: spec.TPRName(),
		},
		Versions: []v1beta1extensions.APIVersion{
			{Name: spec.TPRVersion},
		},
		Description: spec.TPRDescription,
	}
	_, err := c.KubeCli.ExtensionsV1beta1().ThirdPartyResources().Create(tpr)
	if err != nil {
		return err
	}

	return k8sutil.WaitEtcdTPRReady(c.KubeCli.CoreV1().RESTClient(), 3*time.Second, 30*time.Second, c.Namespace)
}
```

Two points:
- it would be really helpful if we provide metrics for k8s api calls
- I doubt the TPR Create() call is taking too long. Let me try. Anothter idea is that Get() might be better.

**Comment by hongchaodeng** [comment]
Seems not caused by TPR create(). Benchmark shows it took ~180ms/op

**Comment by hongchaodeng** [comment]
Seems like it is leader election.

```
time="2017-06-02T16:50:00Z" level=info msg="leader election took 21.58883385s"
```

**Comment by hongchaodeng** [comment]
For solution, I suggest that we will use leader election events as a latch:
```
  FirstSeen	LastSeen	Count	From		SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----		-------------	--------	------		-------
  14s		14s		1	etcd-operator			Normal		LeaderElection	etcd-operator became leader
```

We will add that to:
- starting operator
- upgrading operator
in
- both e2e test and upgrade test to replace WaitTPRReady().

**Comment by hongchaodeng** [comment]
I re-think about this. Let's use endpoints events first to fix that issue. Not to be disruptive for it.

**Comment by hongchaodeng** [comment]
Code references on how to select events:
https://github.com/kubernetes/kubernetes/blob/f5bcd98d5e1fbf7293c3502cc1e6e527a6921825/pkg/printers/internalversion/describe.go#L580-L591

**Comment by xiang90** [comment]
do we know why does election take 20s? was there another operator holding the election so that the new operator needs to wait for it to die?

**Comment by hongchaodeng** [comment]
The other operator has died. There is some setting in the election code that we set around 20 seconds which is the same as scheduler.

**Comment by xiang90** [comment]
> The other operator has died. There is some setting in the election code that we set around 20 seconds which is the same as scheduler.

is it configurable? we probably want to make it shorter to ensure our test can finish faster. but not a big issue.

**Comment by hongchaodeng** [comment]
It's configurable. We can spend time later figuring out what's the best configuration

**Comment by hongchaodeng** [comment]
Closed via #1174
